synapse/types.py

```diff
@@ -421,22 +421,92 @@ class RoomStreamToken:
 
             s0    s1
             |     |
-        [0] V [1] V [2]
+        [0] ▼ [1] ▼ [2]
 
     Tokens can either be a point in the live event stream or a cursor going
     through historic events.
 
-    When traversing the live event stream events are ordered by when they
+    When traversing the live event stream, events are ordered by when they
     arrived at the homeserver.
 
-    When traversing historic events the events are ordered by their depth in
-    the event graph "topological_ordering" and then by when they arrived at the
-    homeserver "stream_ordering".
+    When traversing historic events, events are first ordered by their
+    "depth" ("topological_ordering" in the event graph) and tie-broken by
+    "stream_ordering" (when the event arrived at the homeserver).
+
+    ---
 
     Live tokens start with an "s" followed by the "stream_ordering" id of the
-    event it comes after. Historic tokens start with a "t" followed by the
-    "topological_ordering" id of the event it comes after, followed by "-",
-    followed by the "stream_ordering" id of the event it comes after.
+    event that comes after. The rest of the keys are joined together by
+    underscores in the following order and represent the position of various
+    data.
+
+    ex. `s2633508_17_338_6732159_1082514_541479_274711_265584_1`
+        1. `room_key`: `s2633508` -> `2633508` `stream_ordering`
+        2. `presence_key`: `17`
+        3. `typing_key`: `338`
+        4. `receipt_key`: `6732159`
+        5. `account_data_key`: `1082514`
+        6. `push_rules_key`: `541479`
+        7. `to_device_key`: `274711`
+        8. `device_list_key`: `265584`
+        9. `groups_key`: `1`
```

Comment:
Aren't these values from `StreamToken`s, rather than `RoomStreamToken`? Would it be more appropriate to update that docstring instead?

Reply:
I was just adding to the docstring that seems more like the source of truth. It seems like all of this info should be on one or the other for easier reading and getting the full picture vs splitting it up between.

I don't understand the distinction between `RoomStreamToken` and `StreamToken`? They're intertwined and I can't spot the difference with the `cls` stuff.

Reply:
Using the example above, a `StreamToken` would look like "s2633508_17_338_6732159_1082514_541479_274711_265584_1", with the "s2633508" part being a `RoomStreamToken`.
A `RoomStreamToken` would look like "s2633508", "t426-2633508" or "m56\~2.58\~3.59".

`cls` in a `RoomStreamToken` method looks confusing but it really means "RoomStreamToken". And `cls` in a `StreamToken` method means "StreamToken".

------

I'm not all that keen on having the documentation for `StreamToken` or `RoomStreamToken` live in a different docstring. At the very least, it would be nice if one or the other docstring had some wording along the lines of "See the docstring for `(Room)StreamToken` for a description of/details on <...>"

Reply:
Thank you for the explanation and breakdown @clokep and @squahtx 🙇 

Split the info out to their appropriate classes 👍


---

The naming of these two classes pretty ambiguous. It's all understandable just not immediately clear to me:

 - `RoomStreamToken` is used as the `room_key` in a `StreamToken`. The naming doesn't lend itself to be obvious that a `room_key` is a `RoomStreamToken`
 - `RoomStreamToken` and `StreamToken` includes the word "stream" which immediately jumps my brain to `stream_ordering` and `s999` type tokens. Instead of generic streams of data
 
![](https://user-images.githubusercontent.com/558581/161348194-a3abcc74-1853-4107-a590-113de6cd1f98.gif)

Spit-balling, these seem more obvious although this is a different conversation to have ⏩

  - `RoomStreamToken` -> `RoomPositionKey` (or `RoomStreamPosition`, `RoomPosition`)
  - `StreamToken` -> `PaginationToken` (or `StreamPaginationToken`)